cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/IncapablePlugins.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.mock.batch;
+
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.annotation.Requirements;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.dataset.table.Put;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.etl.api.batch.BatchSink;
+import co.cask.cdap.etl.api.batch.BatchSinkContext;
+import co.cask.cdap.etl.api.batch.BatchSource;
+import co.cask.cdap.etl.api.batch.BatchSourceContext;
+import co.cask.cdap.etl.proto.v2.ETLPlugin;
+
+import java.util.Collections;
+
+/**
+ * A class containing plugins having special requirements for testing failure to start program due to incapability.
+ */
+public class IncapablePlugins {
+  /**
+   * Source which has special requirement
+   */
+  @Plugin(type = BatchSource.PLUGIN_TYPE)
+  @Name(IncapableSource.NAME)
+  @Requirements(IncapableSource.REQUIREMENT)
+  public class IncapableSource extends BatchSource<byte[], Row, StructuredRecord> {
+
+    public static final String NAME = "IncapableSource";
+    public static final String REQUIREMENT = "unicorn";
+
+    @Override
+    public void prepareRun(BatchSourceContext context) throws Exception {
+
+    }
+  }
+
+  /**
+   * Sink which has an requirement which can be meet with another special requirement
+   */
+  @Plugin(type = BatchSink.PLUGIN_TYPE)
+  @Name(IncapableSink.NAME)
+  @Requirements({Requirements.TEPHRA_TX, IncapableSink.REQUIREMENT})
+  public class IncapableSink extends BatchSink<StructuredRecord, byte[], Put> {
+
+    public static final String NAME = "IncapableSink";
+    public static final String REQUIREMENT = "unicorn";
+
+    @Override
+    public void prepareRun(BatchSinkContext context) throws Exception {
+
+    }
+  }
+
+  /**
+   * Get a ETL plugin of the given type
+   *
+   * @param type the type of the plugin which can either be {@link BatchSource#PLUGIN_TYPE} or {@link
+   * BatchSink#PLUGIN_TYPE}
+   *
+   * @return the ETLPlugin of the given type
+   */
+  public static ETLPlugin getPlugin(String type) {
```

Comment:
Avoid this type of pattern where possible, where the caller can easily pass in an invalid value. Each individual class can have its own convenience method to get the ETLPlugin that doesn't require an argument. I also don't think we need this outer class and should make the source and sink their own classes.